packages/vaex-core/vaex/dataframe_protocol.py

```diff
@@ -0,0 +1,688 @@
+    ``vaex.dataframe.DataFrame.__dataframe__`` as objects with the methods and
+    attributes defined on this class.
+    """
+
+    def __init__(self, df: vaex.dataframe.DataFrame, nan_as_null: bool = False, allow_copy: bool = True) -> None:
+        """
+        Constructor - an instance of this (private) class is returned from
+        `vaex.dataframe.DataFrame.__dataframe__`.
+        """
+        self._df = df
+        # ``nan_as_null`` is a keyword intended for the consumer to tell the
+        # producer to overwrite null values in the data with ``NaN`` (or ``NaT``).
+        # This currently has no effect; once support for nullable extension
+        # dtypes is added, this value should be propagated to columns.
+        self._nan_as_null = nan_as_null
+        self._allow_copy = allow_copy
+
+    @property
+    def metadata(self) -> Dict[str, Any]:
+        return {}
+
+    def num_columns(self) -> int:
+        return len(self._df.columns)
+
+    def num_rows(self) -> int:
+        return len(self._df)
+
+    def num_chunks(self) -> int:
+        if isinstance(self.get_column(0)._col.values, pa.ChunkedArray):
+            return self.get_column(0)._col.values.num_chunks
+        else:
+            return 1
+
+    def column_names(self) -> Iterable[str]:
+        return self._df.get_column_names()
+
+    def get_column(self, i: int) -> _VaexColumn:
+        return _VaexColumn(self._df[:, i], allow_copy=self._allow_copy)
+
+    def get_column_by_name(self, name: str) -> _VaexColumn:
+        return _VaexColumn(self._df[name], allow_copy=self._allow_copy)
+
+    def get_columns(self) -> Iterable[_VaexColumn]:
+        return [_VaexColumn(self._df[name], allow_copy=self._allow_copy) for name in self._df.columns]
+
+    def select_columns(self, indices: Sequence[int]) -> "_VaexDataFrame":
+        if not isinstance(indices, collections.Sequence):
+            raise ValueError("`indices` is not a sequence")
+
+        names = []
+        for i in indices:
+            names.append(self._df[:, i].expression)
+
+        return self.select_columns_by_name(names)
+
+    def select_columns_by_name(self, names: Sequence[str]) -> "_VaexDataFrame":
+        if not isinstance(names, collections.Sequence):
+            raise ValueError("`names` is not a sequence")
+
+        return _VaexDataFrame(self._df[names], allow_copy=self._allow_copy)
+
+    def get_chunks(self, n_chunks: Optional[int] = None) -> Iterable["_VaexDataFrame"]:
+        """
+        Return an iterator yielding the chunks.
+        TODO: details on ``n_chunks``
+        """
+        if n_chunks == None:
+            size = self.num_rows()
+            n_chunks = self.num_chunks()
+            i = self._df.evaluate_iterator(self.get_column(0)._col, chunk_size=size // n_chunks)
+            iterator = []
+            for i1, i2, chunk in i:
+                iterator.append(_VaexDataFrame(self._df[i1:i2]))
+            return iterator
```

Comment:
What about sth like:
```suggestion
            for i1, i2, chunk in i:
                yield iterator.append(_VaexDataFrame(vaex.from_dict(chunks))
```
This way consumers can consume the chunks, without us putting everything in memory.